pkg/binding/bindingpolicy-resolution.go

```diff
@@ -297,41 +267,43 @@ func workloadMatchesBindingSpec(bindingSpecWorkload *v1alpha1.DownsyncObjectRefe
 		}
 	}
 
-	if len(objectIdentifierToKeyMap) != clusterScopedObjectsCount+namespacedObjectsCount {
+	if len(objectIdentifiers) != clusterScopedObjectsCount+namespacedObjectsCount {
 		return false
 	}
 
 	// again we can check match by making sure all objects are mapped, since entries are unique and the length is equal.
 	// check cluster-scoped all exist
-	if !bindingClusterScopeIsMapped(bindingSpecWorkload.ClusterScope, objectIdentifierToKeyMap,
-		gvkGvrMapper) {
+	if !bindingClusterScopeIsMapped(bindingSpecWorkload.ClusterScope, objectIdentifiers) {
 		return false
 	}
 
 	// check namespace-scoped all exist
-	return namespaceScopeMatchesBindingSpec(bindingSpecWorkload.NamespaceScope, objectIdentifierToKeyMap,
-		gvkGvrMapper)
+	return namespaceScopeMatchesBindingSpec(bindingSpecWorkload.NamespaceScope, objectIdentifiers)
 }
 
 // bindingClusterScopeIsMapped returns true if the cluster-scope
 // section in the binding spec all exist in the resolution.
 func bindingClusterScopeIsMapped(bindingSpecClusterScope []v1alpha1.ClusterScopeDownsyncObjects,
-	objectIdentifierToKeyMap map[string]*util.Key, gvkGvrMapper util.GvkGvrMapper) bool {
+	objectIdentifiers sets.Set[util.ObjectIdentifier]) bool {
 	for _, clusterScopeDownsyncObjects := range bindingSpecClusterScope {
-		gvr := schema.GroupVersionResource(clusterScopeDownsyncObjects.GroupVersionResource)
-		gvk, found := gvkGvrMapper.GetGvk(gvr)
+		objectNamesSet := sets.New[string](clusterScopeDownsyncObjects.ObjectNames...)
 
-		if !found {
-			return false // if not found then not mapped and not in resolution
-		}
+		// TODO: optimize
+		for objIdentifier := range objectIdentifiers {
+			// skip objects that aren't cluster-scoped
+			if objIdentifier.ObjectName.Namespace != metav1.NamespaceNone {
+				continue
+			}
 
-		for _, objName := range clusterScopeDownsyncObjects.ObjectNames {
-			formattedKey := util.KeyFromGVKandNamespacedName(gvk, types.NamespacedName{
-				Namespace: metav1.NamespaceNone,
-				Name:      objName,
-			})
+			// skip objects that don't match the GVR
+			if objIdentifier.GVK.Group != clusterScopeDownsyncObjects.GroupVersionResource.Group ||
+				objIdentifier.GVK.Version != clusterScopeDownsyncObjects.GroupVersionResource.Version ||
+				objIdentifier.Resource != clusterScopeDownsyncObjects.GroupVersionResource.Resource {
+				continue
+			}
 
-			if _, found := objectIdentifierToKeyMap[formattedKey]; !found {
+			// check if object is in the objectNamesSet
+			if !objectNamesSet.Has(objIdentifier.ObjectName.Name) {
 				return false
 			}
 		}
```

Comment:
This is much more complicated than necessary. Just convert each member of `clusterScopeDownsyncObjects.ObjectNames` to an ObjectIdentifier and test whether `objectIdentifiers.Has` it. Similar comment applies in the namespaced case.

Reply:
The problem is that the Kind is missing from the references in the Binding. To deal with that, I either have to:
1. Have the resolver work on a subtype of ObjectIdentifier that does not contain the Kind (or a new struct type)
2. Implement comparable for ObjectIdentifier to ignore Kind
3. Build sets for both with a type that contains only GVR and ObjectName

I think given that these are the alternatives, I think the implementation is acceptable. If you think otherwise, let me know.

Reply:
Oh, right, you can not use `Set::Has`.